### PR TITLE
Add CI action to test whether feature combinations compile.

### PR DESCRIPTION
**Pull request by nihohit**
This should prevent issues such as https://github.com/redis-rs/redis-rs/issues/1324

### CODE REVIEW

File: redis/Cargo.toml [hunk 1]
```diff
@@ -38,26 +38,26 @@ combine = { version = "4.6", default-features = false, features = ["std"] }
 
 # Only needed for AIO
 bytes = { version = "1", optional = true }
-futures-util = { version = "0.3.15", default-features = false, optional = true }
```
Comment (jaymell) [inline redis/Cargo.toml:41]: Is there a reason to do this? Offhand it seems to me like the patch versions specified in here are useful from a documentation perspective if nothing else. If we don't actually know for sure that the `major.minor.0` versions of these crates will actually compile with the code here, it seems better to not go back and indicate as such now.

File: redis/Cargo.toml [hunk 2]
```diff
@@ -99,7 +99,7 @@ json = ["serde", "serde/derive", "serde_json"]
 cluster = ["crc16", "rand"]
 script = ["sha1_smol"]
 tls-native-tls = ["native-tls"]
-tls-rustls = ["rustls", "rustls/ring", "rustls-native-certs", "rustls-pemfile", "rustls-pki-types"]
+tls-rustls = ["rustls/std", "rustls/ring", "rustls-native-certs", "rustls-pemfile", "rustls-pki-types"]
```
Comment (jaymell) [inline redis/Cargo.toml:102]: Why is this required?

File: redis/src/cluster.rs [hunk 3]
```diff
@@ -238,7 +238,7 @@ where
             connections: RefCell::new(HashMap::new()),
             slots: RefCell::new(SlotMap::new(cluster_params.read_from_replicas)),
             auto_reconnect: RefCell::new(true),
-            read_timeout: RefCell::new(None),
+            read_timeout: RefCell::new(cluster_params.response_timeout),
```
Comment (jaymell) [inline redis/src/cluster.rs:241]: Do we actually need `read_timeout`? Should it just be removed in favor of `cluster_params.response_timeout`?

File: redis/src/cluster.rs [hunk 4]
```diff
@@ -410,7 +410,7 @@ where
         let params = self.cluster_params.clone();
         let info = get_connection_info(node, params)?;
 
-        let mut conn = C::connect(info, None)?;
```
Comment (jaymell) [inline redis/src/cluster.rs:413]: If I'm reading this right, we effectively don't use connection timeout settings in this client? This seems like a potentially impactful to merit a more detailed commit message/possibly a separate PR?

File: redis/src/cluster_client.rs [hunk 5]
```diff
@@ -82,8 +82,8 @@ pub(crate) struct ClusterParams {
     pub(crate) tls: Option<TlsMode>,
     pub(crate) retry_params: RetryParams,
     pub(crate) tls_params: Option<TlsConnParams>,
-    pub(crate) connection_timeout: Duration,
-    pub(crate) response_timeout: Duration,
+    pub(crate) connection_timeout: Option<Duration>,
```
Comment (jaymell) [inline redis/src/cluster_client.rs:85]: Sorry, I vaguely recall we discussed this elsewhere, but remind me why we're doing this? Can we fix the warnings some other way? It seems to me like things are more clear if the defaults are defined here rather than in the connect method.

File: redis/Cargo.toml (hunk 1 above)
Comment (nihohit) [reply redis/Cargo.toml:41]: Doesn't this limit users of the crate? The dependency resolver has to use this specific version of the dependency, and if they have another crate pointing at a different patch version, the dependency is compiled twice?

File: redis/src/cluster.rs (hunk 3 above)
Comment (nihohit) [reply redis/src/cluster.rs:241]: I guess we can remove it - I tried to avoid a larger change.

File: redis/src/cluster.rs (hunk 4 above)
Comment (nihohit) [reply redis/src/cluster.rs:413]: I'll open this as a separate PR, then.

File: redis/src/cluster_client.rs (hunk 5 above)
Comment (nihohit) [reply redis/src/cluster_client.rs:85]: this is done in order to bridge the gap between sync & async clients. ATM the sync client doesn't use these fields, so if we want to use these fields in the sync clients, either we introduce new defaults to the sync client or change the fields to `Option` and set the defaults in the async client.

File: redis/Cargo.toml (hunk 1 above)
Comment (jaymell) [reply redis/Cargo.toml:41]: I don't think it's limiting, though. I think it ensures that only versions >= what's actually been tested are used, which should generally result in fewer old bugs.

File: redis/src/cluster.rs (hunk 3 above)
Comment (jaymell) [reply redis/src/cluster.rs:241]: Ok, avoiding the larger change is probably fine then, or maybe worth considering in the separate PR.

File: redis/src/cluster_client.rs (hunk 5 above)
Comment (jaymell) [reply redis/src/cluster_client.rs:85]: We could also just potentially add an `allow_unused` that's specific to  `!cluster_async` feature?

Comment (nihohit) [reply redis/src/cluster_client.rs:85]: we could, but wouldn't it be better if these fields actually did what they say they do?

Comment (jaymell) [reply redis/src/cluster_client.rs:85]: Yep, I agree. I think adding defaults for the sync client is a good idea.

Comment (nihohit) [reply redis/src/cluster_client.rs:85]: just making sure - you want to apply the same defaults we currently have on the async client to the sync client?

Comment (jaymell) [reply redis/src/cluster_client.rs:85]: Yes, albeit in a separate PR. Alternatively, I'm ok with just allowing the fields to be unused, which although goofy, is effectively what we currently have, right?

Comment (nihohit) [reply redis/src/cluster_client.rs:85]: right.
https://github.com/redis-rs/redis-rs/pull/1333

File: redis/Cargo.toml (hunk 1 above)
Comment (nihohit) [reply redis/Cargo.toml:41]: removed.

File: redis/Cargo.toml (hunk 2 above)
Comment (nihohit) [reply redis/Cargo.toml:102]: removed

Comment (jaymell) [reply redis/Cargo.toml:102]: I was referring to line 102. I still see the change.

Comment (nihohit) [reply redis/Cargo.toml:102]: Sorry, I was wrong. This is needed in order to build the crate with only the `tls-rustls` flag. You can try it locally.

**Review by jaymell — APPROVED** [review]
:sunglasses:
